capsules/src/net/sixlowpan.rs

```diff
@@ -107,21 +231,16 @@ use kernel::hil::time;
 use kernel::hil::time::Frequency;
 use net::frag_utils::Bitmap;
 use net::ieee802154::{PanID, MacAddress, SecurityLevel, KeyId, Header};
-use net::lowpan;
-use net::lowpan::{ContextStore, is_lowpan};
+use net::sixlowpan_compression;
+use net::sixlowpan_compression::{ContextStore, is_lowpan};
 use net::util::{slice_to_u16, u16_to_slice};
 
-// Timer fire rate in seconds
-const TIMER_RATE: usize = 10;
 // Reassembly timeout in seconds
-const FRAG_TIMEOUT: usize = 60;
+const FRAG_TIMEOUT: u32 = 60;
 
-pub trait ReceiveClient {
+pub trait SixlowpanClient {
     fn receive<'a>(&self, buf: &'a [u8], len: u16, result: ReturnCode);
```

Comment:
It still remains to be determined, maybe, if this `'a` lifetime on the buffer makes sense, or if instead the the buffer should be managed by the client. But ok for now. We can change it later if need be

Reply:
The main advantage of this strategy is that we get to pass back an immutable buffer, and be guaranteed that after the callback completes, the RxState can take it's buffer back. More critically, I think that these buffers should *not* be managed by the clients, since dispatching on RxStates is not determined by IP address. In particular, if some client allocates multiple buffers and another client does not, the second client could still receive callbacks (and hence buffers), and simply not return them, exhausting the resources of the first client.

I'm not sure if there is a way to design the upper layers to avoid this problem, but I was thinking that the receive callbacks should all be called in sequence, requiring only a final copy from kernel space to userland.

Reply:
Yep, I think that's sound reasoning. Let's see if there are issues with as we build up the stack.